README
ReachStats 0.8.7
A very feature rich Halo Reach plugin for the IPB platform.
Copyright (c) 2011 ReachStats

ReachStats is released under a Creative Commons Attribution-NonCommerical 3.0 Unported License.
Details of this license can be found here: http://creativecommons.org/licenses/by-nc/3.0/

Bug Reports:  https://github.com/iBotPeaches/ReachStats/issues
Discussion: https://github.com/iBotPeaches/ReachStats
Feature Requests: https://github.com/iBotPeaches/ReachStats/issues

Requirements:
-Bungie API Key ( http://www.bungie.net/Account/reachapikey.aspx )
-Installation of IPB (Invision Power Board)
-Have at least curl, sockets or getfilecontents() 
-GD installed
-TTF (Freetype) installed
-At least php 5.2.11


ReachStats installs as an application into the Invision Power Board platform. Currently supported version is 3.1.4 and will be updated to reflect newer versions as they are released.

Features (see CHANGELOG)
	* Hooks into user control panel for adding of GT
	* Allows entering of Reach API Key directly into ACP
	* Tier system for (Admin, Staff, Trusted Members, Members, Banned)
	* Features small leaderboard page comparing basic stats
	* Gamer profile pages for each user
	* Weapons & Vehicle tables of data per user
	* Pie graph to show win percent
	* List of medals with a count below them
	* Pulls Invasion, Competitive, and Arena
	* Basic stats on Gamer profile page
	* User spartan pulled onto Gamer profile page
	* Direct link for re-caching if permissions allow
	* Shows age of the data on Gamer profile page
	* Dynamic Images for each user
	* Images update based on selected Tier system
	* Shows direct link and a generated forum link
	* Task that runs daily insures that they are all up to date
	* Hooks into userCP popup
	* Sig Options
	* Change some tiny things about images and save your choices
	* Verify gamertags before adding them into system
	* Leaderboard system with a (Gold, Silver, Bronze) ranking.
	
Roadmap 
	* Allow admin to select which playlists are pulled into ReachStats (0.8.5)
	* Abstract Tier Settings out of library file into ACP (0.8.5)
	* Hook into public forum side to show gamer icon (0.8.6)
	* Language Abstraction (0.8.6)
	

CHANGES
== 0.8.7 ==
	Started 3.2 work
	Fixed lang issue
	Fixed wrong API key issue
	Fixed KB articles linking to wrong area
	Fixed templates for 3.2
	Adjusted template on home
	Fixed profile view
	Removed "new profile"

== 0.8.5 ==
	Fixed -1 medal ID problem
	Fixed Rank being returned as null or 0
	Fixed tasks locking themselves
	Fixed random usercpForm error caused by malformed url
	Fixed profileTabs showing up on non-gamer profile forms
	Fixed ReachStuff linking of certain JS files
	Fixed images not loading on certain ways
	Fixed description of Template 2 saying Defeats when it should read Total Games
	
== 0.8.4.b ==
	 Fixed the debug message on profiles

== 0.8.4 ==
	 http://reachstuff.com/community/tracker/issue-79-armory-unlocked armory unlocked. 

== 0.8.3 ==
	 http://reachstuff.com/community/tracker/issue-43-inactive inactive 
	 http://reachstuff.com/community/tracker/issue-49-some-spartans some spartans. 
	 http://reachstuff.com/community/tracker/issue-67-this-registry-getclass $this->registry->getClass 
	 http://reachstuff.com/community/tracker/issue-68-leaderboard-check leaderboard check 
	 http://reachstuff.com/community/tracker/issue-69-no-gt no gt 
	 http://reachstuff.com/community/tracker/issue-71-kd-ratio-is-off kd ratio is off 
	 http://reachstuff.com/community/tracker/issue-72-sig-1-2-signs sig 1 2 % signs 
	 http://reachstuff.com/community/tracker/issue-73-field-types-mysql field types mysql 
	 http://reachstuff.com/community/tracker/issue-74-double-quotes-on-imagepage double quotes on imagePage 
	 http://reachstuff.com/community/tracker/issue-76-commas-challlenges commas challlenges 
	 http://reachstuff.com/community/tracker/issue-77-file-exists-templates file_exists templates 
	 http://reachstuff.com/community/tracker/issue-78-check-w-400-kills-100-games-fails check w/ 400 kills + 100 games fails 

== 0.8.2 ==
	 http://reachstuff.com/community/tracker/issue-44-mass-recache mass recache 
	 http://reachstuff.com/community/tracker/issue-66-preselect-option-for-sig-4 preselect option for sig 4 
	
== 0.8.1 == 
	 http://reachstuff.com/community/tracker/issue-41-queries-image-page queries image page 
	 http://reachstuff.com/community/tracker/issue-48-%20-in-name %20 in Name 
	 http://reachstuff.com/community/tracker/issue-51-non-id-1-on-gamerpage Non ID 1 on gamerpage. 
	 http://reachstuff.com/community/tracker/issue-52-playtime playtime 
	 http://reachstuff.com/community/tracker/issue-53-default-settings default settings 
	 http://reachstuff.com/community/tracker/issue-54-function-findandcheckgtid function findAndCheckGT($id) 
	 http://reachstuff.com/community/tracker/issue-55-compare-page compare page 
	 http://reachstuff.com/community/tracker/issue-56-new-task New task 
	 http://reachstuff.com/community/tracker/issue-57-multiple-medals-check multiple medals check 
	 http://reachstuff.com/community/tracker/issue-59-bsmselect bsmSelect 
	 http://reachstuff.com/community/tracker/issue-60-one-form-per-page ONE FORM PER PAGE. 
	 http://reachstuff.com/community/tracker/issue-61-better-medals better medals 
	 http://reachstuff.com/community/tracker/issue-62-kd kd 
	 http://reachstuff.com/community/tracker/issue-64-balls balls 
	 http://reachstuff.com/community/tracker/issue-65-medals-and-commas medals and commas 

== 0.8.0 ==
	 http://reachstuff.com/community/tracker/issue-36-challenges challenges 
	 http://reachstuff.com/community/tracker/issue-42-1st-green-w-spartan-image 1st green w/ spartan image 
	 http://reachstuff.com/community/tracker/issue-45-timezones-and-challenges timezones and challenges 
	 http://reachstuff.com/community/tracker/issue-46-1st-sig 1st sig. 
	 http://reachstuff.com/community/tracker/issue-47-update-challenges update challenges. 
	 http://reachstuff.com/community/tracker/issue-50-bbcode-imgurl bbcode img/url 

== 0.7.9 ==
	http://reachstuff.com/community/tracker/issue-28-recache-bot recache Bot 
	http://reachstuff.com/community/tracker/issue-29-imagepermissions imagePermissions 
	http://reachstuff.com/community/tracker/issue-30-image-page-new-members Image Page new members 
	http://reachstuff.com/community/tracker/issue-31-totalgames totalGames 
	http://reachstuff.com/community/tracker/issue-32-spartanemblem Spartan/Emblem 
	http://reachstuff.com/community/tracker/issue-33-holy-cow HOLY COW 
	http://reachstuff.com/community/tracker/issue-34-armoury-completion Armoury completion 
	http://reachstuff.com/community/tracker/issue-35-no-killsdeaths No kills/deaths 
	http://reachstuff.com/community/tracker/issue-37-feature-settings feature: settings 
	http://reachstuff.com/community/tracker/issue-38-file-limit file limit 
	http://reachstuff.com/community/tracker/issue-39-explain-pages explain pages 
	http://reachstuff.com/community/tracker/issue-40-auto-check-radio auto-check radio 

== 0.7.8 == (inital release)
	http://reachstuff.com/community/tracker/issue-2-profilerunner profileRunner
	http://reachstuff.com/community/tracker/issue-14-gamerpage gamerpage
	http://reachstuff.com/community/tracker/issue-15-priority-watermarks priority watermarks
	http://reachstuff.com/community/tracker/issue-16-link-to-gamerpage link to gamerpage
	http://reachstuff.com/community/tracker/issue-17-time-between-recaches Time between recaches
	http://reachstuff.com/community/tracker/issue-18-watermark-placement Watermark placement
	http://reachstuff.com/community/tracker/issue-19-bbcode-missing BBCode missing
	http://reachstuff.com/community/tracker/issue-20-remove-junk-from-weap-names Remove junk from weap names
	http://reachstuff.com/community/tracker/issue-21-scalability scalability
	http://reachstuff.com/community/tracker/issue-22-queries queries
	http://reachstuff.com/community/tracker/issue-24-recache-so-soon recache so soon
	http://reachstuff.com/community/tracker/issue-25-watermark-and-donators watermark and donators.
	http://reachstuff.com/community/tracker/issue-26-recache-w-control-0 Recache w/ control = 0


ReachStats/upload/admin/applications_addon/other/reachstat/extensions/usercpForms.php
<?php
/*
   * Copyright (c) 2011 Reach Stats + AUTHORS
   *
   * Permission is hereby granted, free of charge, to any person obtaining a copy
   * of this software and associated documentation files. Please make yourself
   * aware of the license assoicated with this project.
   *
   * THIS SOFTWARE IS PROVIDED "AS IS", WITHOUT ANY WARRANTY OF ANY KIND
   *
   * https://github.com/iBotPeaches/ReachStats
   * bugs: https://github.com/iBotPeaches/ReachStats/issues
   *
   * ~peaches
*/

if (! defined('IN_IPB')) {
    print "<h1>Incorrect access</h1>You cannot access this file directly. If you have recently upgraded, make sure you upgraded all the relevant files.";
    exit();
}

class usercpForms_reachstat  extends public_core_usercp_manualResolver implements interface_usercp {
	public $tab_name 					= "Stats";
    public $ok_message 					= 'Your gamertag exists, and has been saved.';
	public $defaultAreaCode				= 'settings';
    public $hide_form_and_save_button   = false;
    public $uploadFormMax 				= 0;
	public $version						= 32;
	public $tier 						= 0;

    /**
     * Initiate this module
     *
     * @access public
     * @return void
     */
    public function init()
    {
    	/* Set the tab name */
        $this->tab_name = $this->lang->words['reach_manage_gamertag'];

    	/* Class setup */
    	require_once( IPSLib::getAppDir( 'reachstat' ) . '/sources/classReachStats.php' );
    	$this->reach = new reachStats($this->registry );
    	require_once( IPSLib::getAppDir( 'reachstat' ) . '/sources/library.php' );
    	$this->library = new Library( $this->registry );

    	/* Load lang files */
    	$this->registry->getClass('class_localization')->loadLanguageFile( array( 'public_errors', 'public_reachstat' ), 'reachstat');
    }

    /**
     * Return links for this tab
     * You may return an empty array or FALSE to not have
     * any links show in the tab.
     *
     * The links must have 'area=xxxxx'. The rest of the URL
     * is added automatically.
     * 'area' can only be a-z A-Z 0-9 - _
     *
     * @access public
     * @return array array of links
     */
    public function getLinks()
    {
     	$array = array();

        $array[] = array(
        	'url' 		=> 'area=settings',
            'title' 	=> "Gamertag Settings",
            'area' 		=> 'settings',
            'active' 	=> $this->request['tab'] == 'reachstat' && $this->request['area'] == 'settings' ? 1 : 0
            );
    	$array[] = array(
    		'url' 		=>'area=mysigs',
    		'title' 	=> $this->lang->words['sig_tab'],
    		'area'  	=> 'mysigs',
    		'active' 	=> $this->request['tab'] == 'reachstat' && $this->request['area'] == 'mysigs' ? 1 : 0
    		);

        return $array;
    }

    /**
     * Reset the area
     * This is used so that you can reset teh $_AREA, to fix the the left-hand menu highlighting
     * for custom events (e.g. announcements)
     *
     * @access public
     * @param string $ Area
     * @return string Area
     */
    public function resetArea($area)
    {
        return $area;
    }

    /**
     * Run custom event
     *
     * If you pass a 'do' in the URL / post form that is not either:
     * save / save_form or show / show_form then this function is loaded
     * instead. You can return a HTML chunk to be used in the UserCP (the
     * tabs and footer are auto loaded) or redirect to a link.
     *
     * If you are returning HTML, you can use $this->hide_form_and_save_button = 1;
     * to remove the form and save button that is automatically placed there.
     *
     * @access public
     * @param string $ Current 'area' variable (area=xxxx from the URL)
     * @return mixed html or void
     */
    public function runCustomEvent($currentArea)
    {
        // -----------------------------------------
        // INIT
        // -----------------------------------------
        $html = '';

        // -----------------------------------------
        // What to do?
        // -----------------------------------------
        switch ($currentArea) {
            case 'settings':
                $html = $this->_customEvent_reachSettings();
                break;
        	case 'mysigs':
        		$html = $this->_customEvent_sigSettings();
        		break;

        }
        // -----------------------------------------
        // Turn off save button
        // -----------------------------------------
        	$this->hide_form_and_save_button = 1;
        // -----------------------------------------
        // Return
        // -----------------------------------------
        return $html;
    }

    private function _customEvent_reachSettings()
    {

    }

	private function _customEvent_sigSettings()
	{

	}

    /**
     * UserCP Form Show
     *
     * @access public
     * @param string $ Current area as defined by 'get_links'
     * @param array $ Array of member / core_sys_login information (if we're editing)
     * @return string Processed HTML
     */
    public function showForm($current_area, $errors = array())
    {
        // -----------------------------------------
        // Where to go, what to see?
        // -----------------------------------------
        switch ($current_area)
        {
            case 'settings':
                return $this->showReachSettings();
                break;

        	case 'mysigs':
        		return $this->showSigSettings();
        		break;

        }
    }

    /**
     * usercpForms_reachstat::showReachSettings()
     *
     * @return template to show GT
     */
    public function showReachSettings()
    {
    	/* Build the query to grab their gamertag */
            $value = $this->DB->buildAndFetch(array(
                    'select' 		=> 'id,gamertag,ip_address',
                    'from' 			=> 'reachstat',
                    'where' 		=> "id='" . $this->memberData['member_id'] . "'"));

    	/* Get highest Tier */
    	$this->tier = $this->library->determineHighestTier($this->memberData['member_group_id']);

    	/* reset OK message. Previous function changes it */
    	$this->ok_message = 'Your gamertag exists, and has been saved.';

    	/* Remove save button if they added GT */
    	if (($value['gamertag'] != ""))
    	{

    		$this->hide_form_and_save_button = 1;
    	}
    	/* Check if banned */
    	if ($this->tier == 0)
    	{
    		return $this->registry->output->getTemplate('stattrack')->reach_ucp("gt_field", 'Banned Users cannot add GTs');
    	}
    	/* Simple check to see if they have one */
    	if (($value['gamertag'] == ""))
    	{
    		return $this->registry->output->getTemplate('stattrack')->reach_ucp("gt_field", "");
    	}
		else
		{

			/* Convert %20 to space for displaying */
			$value['gamertag'] = $this->reach->unParseGT($value['gamertag']);

			/* head home */
			return $this->registry->output->getTemplate('stattrack')->reach_ucp("gt_field", $value['gamertag']);
		}
    }

	public function showSigSettings()
	{
		/* Add our bsmselect */
		$this->registry->output->addToDocumentHead( 'importcss', $this->settings['board_url'] . '/public/style_css/jquery.bsmselect.css');

		//--------------------------------------------
		// Sig Settings
		//--------------------------------------------
			$value = $this->DB->buildAndFetch(array(
				'select' => 'settings,sigs,data,medals',
				'from'	 => 'reachstat',
				'where'	 => "id='" . $this->memberData['member_id'] . "'"));

		/* set medals in */
		$medals = unserialize($value['medals']);

		//---------------------------------------------
		// Medal Work
		//-----------------------------------------------
		$templateMedals = array ();
		$i = 0;

		/* check if they are medals before doing this */
		if (is_array($medals))
		{
			/* foreach the medals to build a list */
			foreach ($medals as $medal)
			{
				$templateMedals[$i] = ($medal['Name'] . " (" . $medal['value'] . ")");
				$i++;
			}
		}

		/* unset, sort, encode */
		unset($medals);
		asort($templateMedals);
		$templateMedals = json_encode($templateMedals);

		/* check value */
		if ($value['settings'] != null)
		{
			/* undecode */
			$tempValues = unserialize($value['settings']);

			/* dump */
			$radioColor['id'] = intval($tempValues['2']['id']);

			/* check for sp */
			if ($tempValues['2']['sp'] == null)
			{
				$radioColor['sp'] = 0;
			}
			else
			{
				$radioColor['sp'] = $tempValues['2']['sp'];
			}

			//----------------------------------------
			// Sig 1
			//----------------------------------------
			if (($tempValues['1']))
			{
				$i = 0;

				/* Grab each medal */
				foreach ($tempValues['1'] as $medal )
				{
					$pickedMedals[$i] = $medal;
					$i++;
				}
			}

			//-------------------------------------------
			// Sig 5
			//-------------------------------------------
			if (($tempValues['5']))
			{
				/* Set sig 5 color to stored color value */
				$radioColor['color'] = $tempValues['5']['color'];
			}
		}
		else
		{
			/* Set some defaults */
			$radioColor['id'] = 1;
			$radioColor['sp'] = 0;
			$radioColor['color'] = 'aqua'; #Set sig 4 default color
		}

		/* Convert to case names */
		switch($radioColor['id'])
		{
			case 1:
				$radioColor['name'] = '1_green';
				break;
			case 2:
				$radioColor['name'] = '2_blue';
				break;
			case 3:
				$radioColor['name'] = '3_black';
				break;
			case 4:
				$radioColor['name'] = '4_pink';
				break;
			case 5:
				$radioColor['name'] = '5_grey';
				break;
			case 6:
				$radioColor['name'] = '6_red';
				break;
			default:
				$radioColor['name'] = '1_green';
				break;
		}


		//------------------------------------------------------
		// Reload all the possible colors
		//------------------------------------------------------

		/* Bring in the colors */
		$sig5Colors = $this->fillSig5();
		$i = 0;

		/* Find which one they selected */
		foreach ($sig5Colors as $color )
		{
				$sig5[$i]['name'] = $color['name'];
				$sig5[$i]['formal'] = ucfirst(strtolower($color['name'])); #capitilize
				$sig5[$i]['active'] = ($color['name'] == $radioColor['color']) ? 'selected' : null;

			$i++;
		}
		/* clear */
		unset($value);
		unset($tempValues);
		unset($sig5Colors);

		return $this->registry->output->getTemplate('stattrack')->reach_mysigs(json_encode($radioColor), $templateMedals, $pickedMedals, $sig5);
	}

    /**
     * UserCP Form Check
     *
     * @access public
     * @param string $ Current area as defined by 'get_links'
     * @return string Processed HTML
     */
    public function saveForm($current_area)
    {
        // -----------------------------------------
        // Where to go, what to see?
        // -----------------------------------------
        switch ($current_area)
        {
            default:
            case 'settings':
                return $this->saveReachSettings();
                break;

        	case 'mysigs':
        		return $this->saveSigSettings();
        		break;

        }
    }

    /**
     * UserCP Save Form: Settings
     *
     * @access public
     * @return array Errors
     */
    public function saveReachSettings()
    {
    	$id = $this->memberData['member_id'];
		$this->ok_message = 'Your gamertag exists, and has been saved.';

    	/* Get highest Tier */
    	$this->tier = $this->library->determineHighestTier($this->memberData['member_group_id']);

    	/* If Banned Tier, then do not save. */
    	if ($this->tier == 0)
    	{
    		$this->registry->getClass('output')->showError($this->lang->words['tier_banned'], "2007",false,'2007');

    	}

		/* Temp var of their GT */
    	$check = $this->reach->checkGTExists(IPSText::parseCleanValue($this->request['gt_field']),true);

    	$this->DB->freeResult();
    	$this->DB->execute();

    	/* Lets grab their GT and check */
    	$value = $this->DB->buildAndFetch(array(
					'select' 	=> 'gamertag,id,ip_address',
					'from' 		=> 'reachstat',
					'where' 	=> "id='" . intval($id) . "'"));

    	/* Update thier stuff */
    	require_once( IPSLib::getAppDir( 'reachstat' ) . '/sources/classReachStats.php' );
    	$this->reach = new reachStats( $this->registry );
    	require_once( IPSLib::getAppDir( 'reachstat' ) . '/sources/classImageMaking.php' );
    	$this->image = new imageClass( $this->registry );

    	/* Lets see if the GTs are different, meaning they are changing */
    	if (($this->reach->unParseGT($value['gamertag']) != $this->request['gt_field']))
    	{
    		$this->registry->getClass('output')->showError($this->lang->words['gt_changed'], "2024",false,'2024');
    	}
    	//------------------------------------------------
    	// Update Sigs/Data first time
    	//------------------------------------------------

    	/* store default settings */
    	$this->reach->doItAll($this->memberData['member_id']);
    	$this->image->doThemAll($this->memberData['member_id'], $value['gamertag']);

    	/* set default settings */
    	$this->saveSigSettings();

    	/* Check */
    	if (($value['gamertag'] == $this->request['gt_field']))
    	{
    		return false;
    	}
    	else
    	{

    	/* Make sure this GT exists. No morons passing fakes here */
		if ($check == true)
		{
			$this->registry->getClass('output')->showError($this->lang->words['tag_no_exist'], "2000",false,'2000');
		}
    		else
    		{
    			/* set the field */
    			$gamertag = IPSText::parseCleanValue($this->request['gt_field']);

    			/* bi bi */
    			return true;
    		}

			/* If we get down here, something went wrong. Return false */
    		return false;
    	}

    }

	/**
	 * usercpForms_reachstat::saveSigSettings()
	 *
	 * @return
	 */
	public function saveSigSettings()
	{
		/* again :o */
		require_once( IPSLib::getAppDir( 'reachstat' ) . '/sources/classReachStats.php' );
		$this->reach = new reachStats( $this->registry );

		/* Set to 0, just in case they aren't populated */
		$radio['2']['id'] = 0;
		$radio['2']['sp'] = 0;

		/* Grab each radio and then foreach */
		$radio['2']['id'] = $this->request['radioColor']['value'];
		$radio['2']['sp'] = $this->request['spartan_enable'];

		$radio['5']['color'] = $this->request['sig5Colors'];

		/* Check if were empty */
		$i = 0;
		if (!(empty($this->request['medals'])) && count($this->request['medals'] > 3))
		{
			foreach ($this->request['medals'] as  $medal)
			{
				/* Get last occurence of a space, then go from 0 to that space */
				$radio['1'][$i]['Name'] = substr($medal, 0, strrpos($medal,' '));
				$radio['1'][$i]['Id']   = $this->reach->matchMedalToID($radio['1'][$i]['Name']);
				$i++;
			}
			/* Set OK message */
			$this->ok_message = 'Your sig settings have been saved';
		}
		else
		{
			$value = $this->DB->buildAndFetch(array(
				'select' => 'settings',
				'from'	 => 'reachstat',
				'where'	 => "id='" . $this->memberData['member_id']."'"));

			/* set previous settings in */
			$temp = unserialize($value['settings']);
			$radio['1'] = $temp['1'];

			/* tenary just in case they haven't used the 5th sig image */
			$radio['5'] = ($radio['5'] == null ) ? $temp['5'] : $this->request['sig5Colors'];

			/* bi */
			unset($temp);
			unset($value);

		}

		/* Check for too many medal selections */
		if (count($this->request['medals']) > 4)
		{
			/* Set OK message */
			$this->ok_message = 'You have selected too many medals. The max is 4.';
		}

		/* Send to DB */
		$this->DB->update('reachstat', array(
			'id' 		=> intval($this->memberData['member_id']),
			'settings'  => serialize($radio)),
				"id=" . intval($this->memberData['member_id']));


	}

	private function fillSig5()
	{
		$colour['0']['name'] = "aqua";
		$colour['1']['name'] = "black";
		$colour['2']['name'] = "blue";
		$colour['3']['name'] = "brick";
		$colour['4']['name'] = "brown";
		$colour['5']['name'] = "cobalt";
		$colour['6']['name'] = "coral";
		$colour['7']['name'] = "cyan";
		$colour['8']['name'] = "drab";
		$colour['9']['name'] = "forest";
		$colour['10']['name'] = "gold";
		$colour['11']['name'] = "green";
		$colour['12']['name'] = "ice";
		$colour['13']['name'] = "khaki";
		$colour['14']['name'] = "lavender";
		$colour['15']['name'] = "maroon";
		$colour['16']['name'] = "olive";
		$colour['17']['name'] = "orchid";
		$colour['18']['name'] = "pale";
		$colour['19']['name'] = "peach";
		$colour['20']['name'] = "rose";
		$colour['21']['name'] = "rust";
		$colour['22']['name'] = "sage";
		$colour['23']['name'] = "seafoam";
		$colour['24']['name'] = "silver";
		$colour['25']['name'] = "steel";
		$colour['26']['name'] = "tan";
		$colour['27']['name'] = "teal";
		$colour['28']['name'] = "violet";
		$colour['29']['name'] = "white";
		$colour['30']['name'] = "yellow";

		/* bi */
		return $colour;
	}
}
